doc/specs/#140 - ZIP Support.md

```diff
@@ -0,0 +1,90 @@
+- Multiple `NestedInstallerFile` entries can be specified for portable nested installer type to support a suite of portable exes that would all be installed under the same package.
+- If a ZIP `InstallerType` is specified, a `NestedInstallerType` is required to be specified.
+
+## Supported Install Scenarios:
+The initial implementation of this feature will only support the following installation scenarios:
+- Single base installer (exe, msi, msix) contained inside an archive file. 
+- Single or multiple portable exes bundled as a suite inside an archive file.
+
+## ZIP Extraction
+The extraction of ZIPs will be done using Windows Shell APIs. ZIP files can be represented as a [ShellFolder](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ishellfolder) object that can be used to manage the contents of the ZIP file. File contents are represented as [ShellItems](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ishellitem), which can be handled using the methods exposed by the [IFileOperation interface](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ifileoperation). If the `NestedInstallerType` is non-portable, the installer file will be copied to a temporary location and used during execution of the installation flow. If the `NestedInstallerType` is portable, all files will be moved to the appropriate install location and symlinks will be created for each `NestedInstallerFile` entry.
+
+> Since we are utilizing Windows Shell APIs, we need to ensure that we are not invoking a UI during the extraction. This can be done by [setting the operation flag to not display any UI](https://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-ifileoperation-setoperationflags). 
+
+> During implementation, we will also need to ensure that this process can work under SYSTEM context.
+
+## ZIP Threat Detection
+ZIP and other archive file types are known threat vectors for malware in the form of ZIP compression bombs. Two of the most common types of compression bombs are multi-layered (recursive) and single-layered (non-recursive). ZIP bombs rely on a repetition of identical files that have extremely large compression ratios. 
+
+1. **Multi-layered (recursive)**:
+A ZIP file containing multiple layers of nested ZIP files that achieve high compression ratios when decompressed recursively. This technique is often used to bypass compression ratio checks performed by ZIP parsers for a single layer.
+
+2. **Single-Layered (non-recursive)**
+A ZIP bomb that expands fully after a single round of decompression. The extremely high compression ratio of the ZIP bomb is achieved by overlapping files within the ZIP container. 
+
+In order to protect our users from these possible threats, we will need to implement several basic checks to verify whether it is safe to proceed with the contents of the ZIP archive file.

```

Comment:
Do we have a way to prevent these threats on other installer types? For example, imagine I create an EXE with a ZIP bomb embedded and have the EXE decompress it. Would we have the same issues we are trying to avoid here?

Reply:
> Do we have a way to prevent these threats on other installer types? For example, imagine I create an EXE with a ZIP bomb embedded and have the EXE decompress it. Would we have the same issues we are trying to avoid here?

Thats why we have the pipeline validation, isn’t it? If an exe were a zip bomb, it would hang the VM running the validation and trigger an Internal-Error, assuming it made it past the AV scan in the first place

Reply:
That's part of it, but we would like to be able to offer some protection for users accessing third party REST sources as well.

Reply:
That brings up a good point, this scenario would need to additionally scan/verify whether the provided EXE has an embedded zip file, and once that is determined, we could apply the checks here. I do feel like scanning installers (as opposed to scanning a zip file before extracting to obtain an installer) seems to be out of the scope of winget.

Reply:
@ryfu-msft, scanning what `winget`'s default repository provides is certainly not out of its scope, because every Linux distribution's repositories do that, and all of the software that is available via the `msstore` repository has been, too.

Reply:
> @ryfu-msft, scanning what `winget`'s default repository provides is certainly not out of its scope, because every Linux distribution's repositories do that, and all of the software that is available via the `msstore` repository has been, too.
> 
> Additionally, Microsoft's threat-detection services are so brilliant, not utilizing them would be really weird.

Thats not what Ryan was saying at all. The winget-pkgs repository does have all of the packages scanned before they can be added. What Ryan was saying is that private sources and third-party REST sources are outside of the scope of winget in terms of scanning the installer files they provide for security.

You’re also conflating “Package manager” with “Package source”. On Linux especially, apt and snap are package managers, but they do not scan the entire distribution network as you said. Rather, they only fetch the package from the source - and in order to get into the source the package has to be scanned. The difference with apt, snap, and msstore is that they also host the binary files. Winget-pkgs does not host the binaries, instead it hosts the manifests which point to where to download the binaries and how to install them. 

Apt and snap also allow users to host their own private repositories, and similar to winget, there is no guarantee that packages from those repositories are safe. This is because, like winget, apt and snap are just package managers - they get information about the package and have the ability to install, uninstall, and upgrade the package. It is, and almost always has been, the responsibility of the package source, or package repository, to maintain the integrity of their packages. The package managers do very little to ensure security outside of verifying that the installer is the correct installer and that the user has permission to perform the action which is being requested.

Reply:
We will now leverage a zip malware checking library to scan zip files rather than implementing our own malware detection.